rfc-087-use-libraries-where-possible.md

```diff
@@ -0,0 +1,97 @@
+
+## Summary
+
+We should use existing open source libraries wherever possible,
+rather than writing our own code, but we should ensure that those
+libraries are trustworthy as best we can.
+
+## Problem
+
+Writing and maintaining code takes time, and any code is a possible
+source of bugs. When we implement our own solutions for problems
+that exist elsewhere, we may repeat work, increase development time,
+and increase the potential for error.
+
+## Proposal
+
+When we need to add new generic functionality to our code, we have a
+choice of whether to write our own implementation, or use an
+off-the-shelf open source library.
+
+By default, we SHOULD aim to find and reuse existing open source
+libraries that fulfil the requirement rather than writing our own.
+
+Using existing code is preferable in order to get better tested
+code, a lower maintenance burden, and more features than we might
+normally implement ourselves.
+
+Some examples might include:
+
+* Date handling and formatting
+* Language inflection
+* Queueing and background job running
+* Connecting to remote APIs
+
+The need for this will differ across languages and frameworks. For
+instance, Rails provides many capabilities that separate libraries
+would be needed for when working in Node.js. Conversely, frontend
+code may wish to avoid dependencies to keep file sizes low.
+
+Acquiring dependencies is not free, however. Developers MUST check
+them for license compatibility, and SHOULD assess them for quality
+and reliability, so that we are sure they will be a net positive
+addition to our projects.
+
+### License Compatibility
+
+Our code is MIT licensed by default. All dependencies MUST have
+compatible licensing. In particular this means looking out for
+accidental aquisition of a dependency on GPL software (or other
+[copyleft](https://en.wikipedia.org/wiki/Copyleft) licenses that put
+license conditions on derivatives).
+
+You can use tools to check license compatibility in your entire
+dependency tree:
+
+* Ruby: `bundle licenses | sed 's/.*: //g' | sort | uniq`
+* Node.js: `npx license-checker --summary` ([setup instructions](https://www.npmjs.com/package/license-checker))
+
+In general, [permissive licenses](https://en.wikipedia.org/wiki/Permissive_software_license)
+are considered to be MIT compatible, in particular:
+
+* [Apache](https://opensource.org/licenses/Apache-2.0)
+* BSD variants: [3-Clause](https://opensource.org/licenses/BSD-3-Clause),
+  [2-Clause](https://opensource.org/licenses/BSD-2-Clause), or
+  [0BSD](https://opensource.org/licenses/0BSD)
+* [ISC](https://opensource.org/licenses/ISC)
+* [MPL](https://opensource.org/licenses/MPL-2.0)
+* [Unlicense](https://opensource.org/licenses/unlicense)
+
+### Quality and Reliability
+
+We SHOULD only use dependencies that are maintained, trustworthy,
+and of a decent quality. It is nigh-impossible for us to inspect the
+source code of all our dependencies, however, so developer MAY use
+proxy measures to judge this.
```

Comment:
I think it would be useful to give more guidance on determining this.

For example, what signals are we looking for to be confident the dependency is maintained? Activity on issues? Multiple contributors? Active mailing lists or other community?

For proxy measures, what might those be? Is that what you consider Libraries.io? What about whether mainstream projects, like Rails, trust the dependency? Is it OK to implicitly trust any dependency that Rails adds?

Reply:
Yeah, the following paragraphs are intended to lay that out but I could be clearer. Basically libraries.io has a really good list of things it checks. I will reword how I'm pointing to that, I think the order may be a bit wrong.

I also meant to write something about how deep in the tree we go. While the license checkers will do the whole thing, we're not going to check dependencies of dependencies for quality. I would suggest that checking the direct dependency is fine, but deeper is probably unnecessary.

That wouldn't have saved us from the Rails MIME GPL thing, but hardly anything would have, I think, and it's an unreasonable burden to expect to find that sort of thing proactively.

Reply:
I second what Dom's said. I'd like some more guidance on how to go about a manual review. What should someone be looking for (open and unresponded to PRs, PRs closed without response, open issues, test suite and coverage, child dependencies, etc)?

@RobjS Do you have any insight for what you look at for WordPress plugins?

Reply:
We have a command line tool we use when inspecting WordPress plugins: https://git.govpress.com/dxw/pluginscan. It's a blunt tool that searches the code for patterns commonly associated vulnerabilities (e.g. unsanitized user input, manual database queries, network functions that might be sending data to external servers). That helps us focus our code inspection, which is normally time-boxed to an hour or so. We also look for more generic issues like poor coding style (e.g. all the code is a single file made up of functions hundreds of lines long), or unexpectedly large codebases, that might signal an increased likelihood of vulnerabilities being present, even if we haven't spotted them.

Many plugins don't have publicly available git repos, so we don't generally look at things like open PRs & issues. But they do all have a "last updated" date, and a "Tested up to" number, that indicated the latest version of WordPress they've been tested with. I'd expect that to be the latest release (or at least the latest minor version) for any well-maintained plugin.

Reply:
A follow-up question on this. We have https://advisories.dxw.com/ for storing (and, eventually, publishing) the results of our plugin inspections. If we're going to be checking dependencies more generally, in a way that takes a non-trivial amount of time, should we be recording the outcome of those checks somewhere to prevent duplication of effort? If so, where?

Reply:
I've tweaked some things that are relevant to this, see what you think. Might need another pass though.